src/tokens.rs

```diff
@@ -407,8 +407,45 @@ pub fn process_distribute_tokens<T: Client>(
 
     distribute_tokens(client, &mut db, &allocations, args)?;
 
-    let confirmations = update_finalized_transactions(client, &mut db)?;
-    Ok(confirmations)
+    finalize_transactions(client, &mut db, args.dry_run, args.no_wait)
+}
+
+fn finalize_transactions<T: Client>(
+    client: &ThinClient<T>,
+    db: &mut PickleDb,
+    dry_run: bool,
+    no_wait: bool,
+) -> Result<Option<usize>, Error> {
+    let (mut opt_confirmations, mut unconfirmed_transactions) =
+        update_finalized_transactions(client, db)?;
+
+    if no_wait {
+        return Ok(opt_confirmations);
+    }
+
+    let progress_bar = new_spinner_progress_bar();
+
+    while opt_confirmations.is_some() || !unconfirmed_transactions.is_empty() {
+        let confirmations = opt_confirmations.unwrap();
+        progress_bar.set_message(&format!(
+            "[{}/{}] Finalizing transactions",
+            confirmations, 32,
+        ));
+
+        if !dry_run {
+            for transaction in unconfirmed_transactions {
+                client.async_send_transaction(transaction)?;
```

Comment:
This is my least favorite part. We shouldn't need to plumb transactions through the system just to do retries.  This should be managed under the hood by `client.send_transaction()`.  The retries can be done by a client thread or by the server.  Added #29 for tracking.